lib/upsert/merge_function/PG_Connection.rb

```diff
@@ -6,7 +6,70 @@ class MergeFunction
     class PG_Connection < MergeFunction
       include Postgresql
 
+      attr_reader :quoted_setter_names
+      attr_reader :quoted_selector_names
+
+      def self.use_pg_upsert_cache
+        @use_pg_upsert_cache ||= {}
+      end
+
+      def initialize(controller, *args)
+        super
+        @quoted_setter_names = setter_keys.map { |k| connection.quote_ident k }
+        @quoted_selector_names = selector_keys.map { |k| connection.quote_ident k }
+      end
+
       def execute(row)
+        use_pg_upsert? ? pg_upsert(row) : pg_function(row)
+      end
+
+      def use_pg_upsert?
+        return if @use_native_upsert == false
+        # We need to memoize this way because we're storing a boolean
+        return self.class.use_pg_upsert_cache[__id__] if self.class.use_pg_upsert_cache.key?(__id__)
+
+        self.class.use_pg_upsert_cache[__id__] = begin
+          version = controller.connection.execute("SHOW server_version").getvalue(0, 0)
+          version_number = version[0..2].split('.').join('').to_i
+
+          matching_constraint = false
+          schema_query = controller.connection.execute("SELECT array_agg(column_name::text) FROM information_schema.constraint_column_usage WHERE table_name = $1 GROUP BY table_catalog, table_name, constraint_name", [table_name])
+          type_map = PG::TypeMapByColumn.new([PG::TextDecoder::Array.new])
+          schema_query.type_map = type_map
+
+          schema_query.values.each do |row|
```

Comment:
I think something's a bit wacky with the review.  This one's definitely resolved: https://github.com/seamusabshere/upsert/pull/79/files#diff-d7104b6a97c3433dcd1a8bceffb9584dR41